### PR TITLE
Add And layer

### PR DESCRIPTION
**Pull request by Vimalakirti**

### CODE REVIEW

File: src/onnx.rs [hunk 1]
```diff
@@ -125,6 +125,7 @@ fn get_local_graph(
 ) -> (Graph, Vec<Vec<usize>>) {
   match op {
     "Add" => Ok(AddLayer::graph(&input_shapes, &node_constants, &node_attributes)),
+    "And" => Ok(MulLayer::graph(&input_shapes, &node_constants, &node_attributes)), // use MulLayer for And
```
Comment (ddkang) [inline src/onnx.rs]: Ditto: Does this require that the inputs are in {0, 1}? Is this implicitly implied by the inputs?

Comment (Vimalakirti) [reply src/onnx.rs]: Probably the easiest way is we use a basicblock to perform the boolean check during model compilation.

File: src/basic_block/bool_check.rs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+use super::{BasicBlock, Data, DataEnc, PairingCheck, ProveVerifyCache, SRS};
```
Comment (ddkang) [inline src/basic_block/bool_check.rs]: Is this secure? If not, please make an issue and note it in a comment

Comment (Vimalakirti) [reply src/basic_block/bool_check.rs]: I think this depends on how we define security. We say a layer is secure if:
(1) It correctly performs the defined functionalities, or
(2) It correctly performs the defined functionalities and checks the correctness of inputs.

According to definition (1), the previous layer should be responsible for the correctness of the inputs to the current layer. If all previous layers are secure, then the correctness of inputs is delegated, ensuring the inputs of the entire model are correct.

According to definition (2), each layer should verify its input format.

I added new `prove` and `verify` for Boolean checks considering our consensus might be (2).

Comment (Vimalakirti) [reply src/basic_block/bool_check.rs]: Added an issue #83 because we do not need bool check(s) after a Cast layer.

**Comment by liliatangxy** [comment]
lgtm minus the blinding comment in https://github.com/uiuc-kang-lab/zk-torch/pull/74

**Comment by Vimalakirti** [comment]
> lgtm minus the blinding comment in #74

Thanks, comments added.

**Comment by ddkang** [comment]
Now that we know the fix, why can't we fix it now?

**Comment by Vimalakirti** [comment]
Fixed the opening issue. Now it is secure.

**Comment by Vimalakirti** [comment]
From now on, please ignore #72 and #74. Let us focus on And and get it merged first.

**Review by liliatangxy — DISMISSED** [review]
Blinding looks good

**Review by ddkang — APPROVED** [review]

**Pull request merged**